gtk4/src/subclass/widget.rs

```diff
@@ -895,6 +896,118 @@ pub unsafe trait WidgetClassSubclassExt: ClassStruct {
         }
     }
 
+    /*
+    fn install_action<F: Fn(&Widget, &str, Option<&glib::Variant>) + 'static>(
+        &mut self,
+        action_name: &str,
+        parameter_type: Option<&str>,
+        f: F,
+    ) {
+        unsafe {
+            unsafe extern "C" fn activate_trampoline<
+                F: Fn(&Widget, &str, Option<&glib::Variant>) + 'static,
+            >(
+                this: *mut ffi::GtkWidget,
+                action: *const libc::c_char,
+                parameter: *mut glib::ffi::GVariant,
+            ) {
+                f(
```

Comment:
This is almost working, but I have no idea how can i pass the callback to here as there's no way to pass user_data to the callback, any ideas?

Reply:
Well this sucks, it's broken API. The only thing you can do here is to install a `fn(*mut gtk::ffi::GtkWidget, *const c_char, *mut glib::ffi::GVariant)` here.

This needs some new API added to GTK that takes a `user_data`, and if they care about dynamic bindings also a `GDestroyNotify` (that would never be called).

The annotation of the C function is also completely wrong: ` * @activate: (scope call): callback to use when the action is activated`. This is not `scope call`, it's a callback that has to be forever valid.

tl;dr: Fix it first in C by adding some new API there. This function as-is is completely useless for bindings.

Reply:
See also https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/1732#note_989762 . I'm surprised that this wasn't noticed as part of the review of that MR

Reply:
You could probably work around it by storing a hash table in the `interface_data` in `ObjectSubclass::type_data()` here that maps from the `action_name` (second parameter here and in the callback) to the `user_data` (the `Box<dyn Fn>` closure).

See [here](https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/blob/3428e5c0c753f7eae8a60dcae7cbf30c9828cd88/gstreamer/src/subclass/uri_handler.rs#L26-34) and [here](https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/blob/3428e5c0c753f7eae8a60dcae7cbf30c9828cd88/gstreamer/src/subclass/uri_handler.rs#L53-54) for some usage of that.

It's all rather awkward though.

@ids1024 FYI, another usage of the `type_data()` that is potentially accessed a lot

Reply:
I have went ahead and opened an issue at https://gitlab.gnome.org/GNOME/gtk/-/issues/3498 for now

Reply:
I went ahead and disabled this for now; i will try adding a gclosure alternative to gtk first & use it here in the future

Reply:
Can you create an issue about that? Seems rather fundamental :)

Reply:
There's an issue already to track all the widget class methods, see #111

Reply:
Going back to this, I'm a bit stuck on how to properly do thing. Here's what I have currently more or less 
```rust
fn install_action<F: Fn(&Widget, &str, Option<&Variant>) + 'static>(
        &mut self,
        action_name: &str,
        parameter_type: Option<&str>,
        activate: F,
    ) {
        // We store the activate callbacks in a HashMap<action_name, activate>
        // so that we can retrieve f later on the activate_trampoline call
        let mut data = <Self::Type as ObjectSubclass>::type_data();
        let data = data.as_mut();

        #[derive(Debug)]
        struct Actions(pub(crate) HashMap<String, glib::ffi::gpointer>);
        unsafe impl Sync for Actions {}
        unsafe impl Send for Actions {}

        let f: Box_<F> = Box_::new(activate);

        let actions = data
            .get_class_data_mut::<Actions>(Widget::static_type())
            .unwrap();
        actions.0.insert(action_name.to_string(), Box::into_raw(f));

        unsafe extern "C" fn activate_trampoline<F>(
            this: *mut ffi::GtkWidget,
            action_name: *const libc::c_char,
            parameter: *mut glib::ffi::GVariant,
        ) {
            let action_name = GString::from_glib_borrow(action_name);

            let mut data = <Self::Type as ObjectSubclass>::type_data();
            let data = data.as_mut();

            let actions = data
                .get_class_data::<Actions>(Widget::static_type())
                .unwrap();
            let activate_callback = actions.0.get(&action_name.to_string()).unwrap();

            let f: &F = &*(activate_callback as *const F);
            f(
                &from_glib_borrow(this),
                &action_name,
                Option::<Variant>::from_glib_borrow(parameter)
                    .as_ref()
                    .as_ref(),
            )
        }

        unsafe {
            let widget_class = self as *mut _ as *mut ffi::GtkWidgetClass;
            ffi::gtk_widget_class_install_action(
                widget_class,
                action_name.to_glib_none().0,
                parameter_type.to_glib_none().0,
                Some(transmute::<
                    _,
                    unsafe extern "C" fn(*mut ffi::GtkWidget, *const i8, *mut glib::ffi::GVariant),
                >(activate_trampoline::<F> as *const ())),
            );
        }
    }
```

The issue I'm having currently is how can I access the TypeData inside the activate_tramploline as `Self` isn't available in that scope
```rust
            let mut data = <Self::Type as ObjectSubclass>::type_data();
            let data = data.as_mut();
```
Any ideas? :)

Reply:
Oh, maybe I just need to cast `this`? :)

Reply:
Pass `Self` as a second type parameter to the trampoline. `activate_trampoline<F, S>` :)

That seems like a good idea anyway, then you can use `Self::Type` instead of `gtk::Widget` as first parameter of the callback.